torch/csrc/autograd/python_function.cpp

```diff
@@ -212,30 +212,47 @@ static int THPFunction_traverse(THPFunction* self, visitproc visit, void* arg) {
   //
   // TODO: I'm not really sure if we're actually obligated to traverse PyObject
   // that is stored in PyNode, since we don't really own that C++ object.
-  if (auto cdata = self->cdata.lock()) {
-    for (const auto& hook : cdata->tensor_pre_hooks()) {
-      if (auto pyhook = dynamic_cast<PyFunctionTensorPreHook*>(hook.get())) {
-        Py_VISIT(pyhook->dict);
+  // Take care to keep the call to .lock() in a scope, so that we release our
+  // shared pointer before traversing any saved variables.
+  // This edge case can come up if any saved variables (indirectly) hold
+  // a strong reference to cdata (our PyNode).
+  {
+    if (auto cdata = self->cdata.lock()) {
+      for (const auto& hook : cdata->tensor_pre_hooks()) {
+        if (auto pyhook = dynamic_cast<PyFunctionTensorPreHook*>(hook.get())) {
+          Py_VISIT(pyhook->dict);
+        }
       }
-    }
-    // See NOTE [retains_grad_hook PyObject traversal]
-    for (const auto& pair : cdata->retains_grad_hooks()) {
-      if (auto pyhook =
-              dynamic_cast<PyFunctionTensorPreHook*>(pair.second.get())) {
-        Py_VISIT(pyhook->dict);
+      // See NOTE [retains_grad_hook PyObject traversal]
+      for (const auto& pair : cdata->retains_grad_hooks()) {
+        if (auto pyhook =
+                dynamic_cast<PyFunctionTensorPreHook*>(pair.second.get())) {
+          Py_VISIT(pyhook->dict);
+        }
       }
-    }
-    for (const auto& hook : cdata->pre_hooks()) {
-      if (auto pyhook = dynamic_cast<PyFunctionPreHook*>(hook.get())) {
-        Py_VISIT(pyhook->dict);
+      for (const auto& hook : cdata->pre_hooks()) {
+        if (auto pyhook = dynamic_cast<PyFunctionPreHook*>(hook.get())) {
+          Py_VISIT(pyhook->dict);
+        }
       }
-    }
-    for (const auto& hook : cdata->post_hooks()) {
-      if (auto pyhook = dynamic_cast<PyFunctionPostHook*>(hook.get())) {
-        Py_VISIT(pyhook->dict);
+      for (const auto& hook : cdata->post_hooks()) {
+        if (auto pyhook = dynamic_cast<PyFunctionPostHook*>(hook.get())) {
+          Py_VISIT(pyhook->dict);
+        }
       }
     }
   }
+  for (const auto& saved_variable : self->saved_variables) {
+    // Careful not to go through unpack(), which does a lot of extra things and
+    // might create a fresh tensor.
+    const auto& saved_tensor = saved_variable._unsafe_data();
+    if (!saved_tensor.defined()) {
+      continue;
+    }
+    if (saved_tensor.use_count() == 1) {
+      THPVariable_cpp_tensor_visit(saved_tensor, visit, arg);
+    }
```

Comment:
It's not too difficult to contrive an example where the saved tensor has a use count higher than 1,
```python
ctx.save_for_backward(c, c)
```

Or for a less contrived example, you could have a second view:
```python
view0 = result.view(b.shape[0], -1)
view1 = view0.view(-1)
ctx.save_for_backward(view0, view1)
```

Would it make more sense to deprecate/warn on saving a view of the result, and teach the AOTAutograd graph partitioner to put these view ops into the backward graph?

Reply:
Or another way to break the cycle would be for `Function.apply()` to return an alias of `b` in this case. That way the alias owns the `Function`, and both the alias and the `Function` own the original tensor `b`.  Though I guess it might be surprising if the user relies on a tensor not being a view (does this happen?).

Reply:
Sorry for the late reply @peterbell10!

You're right, that:
(1) it looks like this logic needs some more beefing up to handle dupes
(2) we could teach AOTAutograd to partition better to avoid this situation.

In fact, we have an even hackier fix in today: we [manually break](https://github.com/pytorch/pytorch/blob/0da89664cc21b14b84c5d438d358278699f8a51e/torch/_functorch/aot_autograd.py#L2224) the ref cycle by detaching saved tensors in this situation (which is wrong if we want to double backward, but double backward is already explicitly banned in aot autograd).

This PR is/was meant as more of a root-cause fix (although it's still broken. Alban also spent more time staring at it and fixed some issues, but it's still failing CI).

> Would it make more sense to deprecate/warn on saving a view of the result, and teach the AOTAutograd graph partitioner to put these view ops into the backward graph?

I think you're right - although autograd.Function is pretty widely used (not sure what the BC risk of that deprecation is).

Reply:
@bdhirsh and what about returning an alias? Do you know if there are BC implications?